### PR TITLE
Digeq 112 navigation

### PR DESCRIPTION
**Pull request by weapdiv-david**
**What**

_NOTE:_ This builds upon the previous Pull Request for DIGEQ-54

_Fixes issue DIGEQ-160 Runner: Checkbox states not restored_

Also...

1) Creates the summary response page complete with per-section validation status
2) Dynamically removes links from the history based and routing
3) Dynamically removes responses from summary view based on branching and routing
4) Implements unique identifiers for each question 
5) Persists and retores state of previously completed sections that may be affected by branching and routing
6) Dynamically creates a unique session id where one is missing
7) Dynamically updates progress indicators based on branching

**How to test**

There is a lot of functionality in this PR, and a large amount of refactoring work has been done.  In order to test effectively, it is best to view a debug questionnaire and experiment with completing the questionnaire.  **There is a branching rule on the favourite film: Answering "The Phantom Menace" will cause section two to be skipped.**

1) Visit http://localhost:5000/questionnaire/1?debug=true
2) You will be redirected to a new session and the start page
3) Begin filling in the questionaire
4) Hitting Save and Continue will cause the current page to be validated and enable it's appearance in the history
5) You can branch forward, click backwards and modify your responses at any time.
6) The final submission step after the summary page **has not** been implemented yet

**Who can test**

Anybody except @weapdiv-david and @warren-methods

### CODE REVIEW

File: branching.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class Condition(object):
+    def __init__(self, *args):
+        if len(args) == 3:
+            self.target = args[0]
+            self.trigger = args[1]
+            self.state = args[2]
+        elif len(args) == 2:
+            self.trigger = args[0]
+            self.state = args[1]
+
```
Comment (LJBabbage) [inline branching.py:10]: Unused Function

File: branching.py [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
+        elif len(args) == 2:
+            self.trigger = args[0]
+            self.state = args[1]
 
 
 class JumpTo(Condition):
     def __init__(self, target, trigger, state):
         super(JumpTo, self).__init__(target, trigger, state)
+
+class SkipCondition(Condition):
```
Comment (LJBabbage) [inline branching.py]: Pep8 issue

File: branching.py [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
+            self.state = args[2]
+        elif len(args) == 2:
+            self.trigger = args[0]
+            self.state = args[1]
 
 
 class JumpTo(Condition):
     def __init__(self, target, trigger, state):
         super(JumpTo, self).__init__(target, trigger, state)
+
```
Comment (LJBabbage) [inline branching.py]: Pep8 issue

File: questionnaireManager.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 from questions import Question
 import json
-
+from branching import SkipCondition
```
Comment (LJBabbage) [inline questionnaireManager.py]: Pep8 issue with line spaces

File: branching.py [hunk 5]
```diff
@@ -1,10 +1,19 @@
 class Condition(object):
     def __init__(self, *args):
-        self.target = args[0]
```
Comment (LJBabbage) [inline branching.py:3]: The wording confused me a bit, would prefer target, question_reference and condition.

File: questionnaireManager.py [hunk 6]
```diff
@@ -21,24 +23,51 @@ def _load_questionnaire_data(self, questionnaire_data):
...
+                question._reference = 'q' + str(index)
                 
             self._add_question(question)
 
+    def _ensure_valid_routing(self):
+        for question_index, question in enumerate(self.questions):
+            if question.has_branch_conditions():
+                target_questions = {}
+                for condition in question.get_branch_conditions():
+                    # evaluate the condition
```
Comment (LJBabbage) [inline questionnaireManager.py:40]: Would prefer not to use a nested function

question_ref = self.get_question_by_reference(condition.target)
target_question[question_ref]=SkipCondition(condition.trigger, condition.state)

File: branching.py [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
+        if len(args) == 3:
+            self.target = args[0]
+            self.trigger = args[1]
+            self.state = args[2]
+        elif len(args) == 2:
+            self.trigger = args[0]
+            self.state = args[1]
 
 
 class JumpTo(Condition):
```
Comment (LJBabbage) [inline branching.py:12]: A bit picky, but JumpToQuestion would be a more helpful name

File: branching.py [hunk 8]
```diff
@@ -1,10 +1,19 @@
 class Condition(object):
```
Comment (LJBabbage) [inline branching.py]: Whole class needs some comments

File: questionnaireManager.py [hunk 9]
```diff
@@ -117,29 +159,61 @@ def get_next_question(self, response):
...
             index += 1
 
     # will use in template
     def get_question_by_reference(self, reference):
-        for question in self.questions:
-            if question.reference == reference:
-                return question
 
-        return None
+        address_parts = reference.split(':')
```
Comment (LJBabbage) [inline questionnaireManager.py:186]: I understand why you have done this, due to our current schema. However this should be refactored in the beta, Sections and questions should not be represented the same way and therefore this shouldn't be needed

File: questions.py [hunk 10]
```diff
@@ -177,3 +209,30 @@ def get_branch_target(self, response):
...
+                    conditions.append(child_condition)
+
+        return conditions
+
+    def has_branch_conditions(self):
+        return len(self.get_branch_conditions()) > 0
+
+    def get_question_by_reference(self, reference):
+
+        address_parts = reference.split(':')
```
Comment (LJBabbage) [inline questions.py:229]: Same as before, needs refactoring in the beta

File: srunner.py [hunk 11]
```diff
@@ -107,7 +129,13 @@ def questionnaire_viewer(questionnaire_id, quest_session_id=None):
             user_responses = {}
             for key in request.form.keys():
                 if key != 'next' and key != 'start':
-                    user_responses[key] = request.form[key]
```
Comment (LJBabbage) [inline srunner.py:110]: Odd looking if statement ,specially if it can't be -1

File: branching.py (hunk 5 above)
Comment (weapdiv-david) [inline branching.py:3]: question_reference ties it to a question, while this might be used with sections as well

File: branching.py (hunk 7 above)
Comment (weapdiv-david) [inline branching.py:12]: This is the base class for all Conditions, of which JumpTo(target, trigger, state) and SkipCondition(trigger, state) are subclasses.  This base class allows conditions to be handled in roughly the same manner by the caller, while hiding implementation specific details.  At this stage we only have two types of condition, but there may be more needed when we put this in front of real users.

**Pull request merged**